conans/test/functional/test_third_party_patch_flow.py

```diff
@@ -0,0 +1,199 @@
+import os
+import textwrap
+
+from conans.test.utils.scm import create_local_git_repo
+from conans.test.utils.test_files import temp_folder
+from conans.test.utils.tools import TestClient
+from conans.util.files import mkdir
+
+
+def test_third_party_patch_flow():
+    """ this test emulates the work of a developer contributing recipes to ConanCenter, and having
+    to do multiple patches to the original library source code:
+    - Everything is local, not messing with the cache
+    - Using layout() to define location of things
+    """
+    conanfile = textwrap.dedent(r"""
+        import os
+        from conans import ConanFile, load
+        from conans.tools import save
+        from conan.tools.files import apply_conandata_patches
+
+        class Pkg(ConanFile):
+            name = "mypkg"
+            version = "1.0"
+            exports_sources = "*"
+
+            def layout(self):
+                self.folders.source = "src"
+                self.folders.build = "build"
+
+            def source(self):
+                # emulate a download from web site
+                save("myfile.cpp", "mistake1\nsomething\nmistake2\nmistake3\nsome\n")
+                apply_conandata_patches(self)
+
+            def build(self):
+                for i in (1, 2, 3):
+                    if "mistake{}".format(i) in load(os.path.join(self.source_folder, "myfile.cpp")):
+                        raise Exception("MISTAKE{} BUILD!".format(i))
+        """)
+
+    client = TestClient()
+    client.save({"conanfile.py": conanfile,
+                 "conandata.yml": ""})
+    client.run("install .")
```

Comment:
I miss where is layout defined here, there is no command-line argument like `--layout` or individual directories like `-if`/`-bf`/`-sf`/`-pf`.
every command just passes an unnamed positional argument `.`, what are these directories in terms of layout?
as I can see, layout is defined inside the recipe, why not outside of it?
in these case, client seems to be required to have a knowledge that this particular recipe uses `src` directory as a source folder.
how should I write my commands below (lines 55-56) so that they work on arbitrary recipes (the one which may define `self.folder.source = "non-standard"`?

Reply:
Please, you should take a look at least at the highlight features we release, the layout feature has some time now. This PR is not about discussing the layouts feature.